File: ARC MVP/EPI/android/app/src/main/kotlin/com/example/my_app/QwenBridge.kt
package com.example.my_app

import android.app.Activity
import android.app.ActivityManager
import android.content.Context
import android.os.Build
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*
import java.io.File
import kotlin.random.Random

// Load the native library
init {
    System.loadLibrary("llama_cpp_wrapper")
}

// Native function declarations
external fun llamaInit(modelPath: String): Int
external fun llamaGenerate(prompt: String, temperature: Float, topP: Float, maxTokens: Int): String?
external fun llamaCleanup()
external fun llamaIsLoaded(): Int

/**
 * QwenBridge - Native Android bridge for LUMARA Qwen models
 * 
 * Provides unified interface for:
 * - Qwen3-4B-Instruct / 1.7B (Chat)
 * - Qwen2.5-VL-3B / Qwen2-VL-2B (Vision)
 * - Qwen3-Embedding-0.6B (Embeddings)
 * 
 * Runtime support: llama.cpp (GGUF) and MLC LLM
 */
class QwenBridge : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    
    // Model state tracking
    private var chatModelLoaded = false
    private var visionModelLoaded = false
    private var embeddingModelLoaded = false
    
    // Current runtime configuration
    private var currentRuntime = "llamacpp"
    private var temperature: Float = 0.6f
    private var topP: Float = 0.9f
    private var maxTokens: Int = 256
    
    companion object {
        private const val CHANNEL = "lumara_native"
        private const val TAG = "QwenBridge"
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            // Chat model methods
            "initChatModel" -> initChatModel(call, result)
            "qwenText" -> qwenText(call, result)
            
            // Vision model methods  
            "initVisionModel" -> initVisionModel(call, result)
            "qwenVision" -> qwenVision(call, result)
            
            // Embedding model methods
            "initEmbeddingModel" -> initEmbeddingModel(call, result)
            "embedText" -> embedText(call, result)
            "embedTextBatch" -> embedTextBatch(call, result)
            
            // Device capabilities
            "getDeviceCapabilities" -> getDeviceCapabilities(result)
            
            // Model management
            "isModelReady" -> isModelReady(call, result)
            "getModelLoadingProgress" -> getModelLoadingProgress(call, result)
            
            // Runtime management
            "switchRuntime" -> switchRuntime(call, result)
            "getRuntimeInfo" -> getRuntimeInfo(result)
            
            "dispose" -> dispose(result)
            
            else -> result.notImplemented()
        }
    }

    // MARK: - Chat Model Methods

    private fun initChatModel(call: MethodCall, result: Result) {
        val modelPath = call.argument<String>("modelPath")
        val temperature = call.argument<Double>("temperature") ?: 0.6
        val topP = call.argument<Double>("top_p") ?: 0.9
        val maxTokens = call.argument<Int>("max_tokens") ?: 256

        if (modelPath == null) {
            result.success(false)
            return
        }

        println("$TAG: Initializing chat model")
        println("  Model path: $modelPath")
        println("  Temperature: $temperature")
        println("  Top-P: $topP")
        println("  Max tokens: $maxTokens")

        // Store parameters for later use
        this.temperature = temperature.toFloat()
        this.topP = topP.toFloat()
        this.maxTokens = maxTokens

        // Initialize llama.cpp model
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val success = llamaInit(modelPath)
                
                withContext(Dispatchers.Main) {
                    if (success == 1) {
                        chatModelLoaded = true
                        println("$TAG: Chat model loaded successfully")
                        result.success(true)
                    } else {
                        println("$TAG: Failed to load chat model")
                        result.success(false)
                    }
                }
            } catch (e: Exception) {
                println("$TAG: Error initializing model: ${e.message}")
                withContext(Dispatchers.Main) {
                    result.success(false)
                }
            }
        }
    }

    private fun qwenText(call: MethodCall, result: Result) {
        val prompt = call.argument<String>("prompt")
        
        if (prompt == null) {
            result.success("")
            return
        }

        if (!chatModelLoaded) {
            println("$TAG: Chat model not loaded")
            result.success("")
            return
        }

        println("$TAG: Generating text for prompt: ${prompt.take(50)}...")

        // Generate response using llama.cpp
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val response = llamaGenerate(prompt, temperature, topP, maxTokens)
                
                withContext(Dispatchers.Main) {
                    if (response != null) {
                        println("$TAG: Generated response: ${response.take(100)}...")
                        result.success(response)
                    } else {
                        println("$TAG: Failed to generate response")
                        result.success("")
                    }
                }
            } catch (e: Exception) {
                println("$TAG: Error generating response: ${e.message}")
                withContext(Dispatchers.Main) {
                    result.success("")
                }
            }
        }
    }

    // MARK: - Vision Model Methods

    private fun initVisionModel(call: MethodCall, result: Result) {
        val modelPath = call.argument<String>("modelPath")
        
        if (modelPath == null) {
            result.success(false)
            return
        }

        println("$TAG: Initializing vision model at $modelPath")

        // TODO: Initialize actual Qwen-VL model
        CoroutineScope(Dispatchers.IO).launch {
            delay(1500) // Simulate VLM loading time
            visionModelLoaded = true
            withContext(Dispatchers.Main) {
                result.success(true)
            }
        }
    }

    private fun qwenVision(call: MethodCall, result: Result) {
        val prompt = call.argument<String>("prompt")
        val imageJpeg = call.argument<ByteArray>("imageJpeg")
        
        if (prompt == null || imageJpeg == null) {
            result.success("")
            return
        }

        if (!visionModelLoaded) {
            println("$TAG: Vision model not loaded")
            result.success("")
            return
        }

        println("$TAG: Analyzing image (${imageJpeg.size} bytes) with prompt: $prompt")

        // TODO: Call actual Qwen-VL model inference
        CoroutineScope(Dispatchers.IO).launch {
            delay(1000) // Simulate vision inference time
            
            val simulatedVisionResponse = """
                I can see the image you've shared. This appears to be a photo related to your journal entry or personal experience.
                
                Your question: "$prompt"
                
                *This is a stub response. The actual Qwen2.5-VL model will provide detailed image analysis and answer questions about visual content once the integration is complete.*
            """.trimIndent()
            
            withContext(Dispatchers.Main) {
                result.success(simulatedVisionResponse)
            }
        }
    }

    // MARK: - Embedding Model Methods

    private fun initEmbeddingModel(call: MethodCall, result: Result) {
        val modelPath = call.argument<String>("modelPath")
        
        if (modelPath == null) {
            result.success(false)
            return
        }

        println("$TAG: Initializing embedding model at $modelPath")

        // TODO: Initialize actual Qwen3-Embedding model
        CoroutineScope(Dispatchers.IO).launch {
            delay(500) // Simulate embedding model loading
            embeddingModelLoaded = true
            withContext(Dispatchers.Main) {
                result.success(true)
            }
        }
    }

    private fun embedText(call: MethodCall, result: Result) {
        val text = call.argument<String>("text")
        
        if (text == null) {
            result.success(listOf<Double>())
            return
        }

        if (!embeddingModelLoaded) {
            println("$TAG: Embedding model not loaded")
            result.success(listOf<Double>())
            return
        }

        println("$TAG: Generating embeddings for text: ${text.take(50)}...")

        // TODO: Generate actual embeddings with Qwen3-Embedding
        CoroutineScope(Dispatchers.IO).launch {
            delay(100) // Simulate embedding generation
            
            // Return simulated 512-dimensional embeddings
            val simulatedEmbeddings = (0 until 512).map { Random.nextDouble(-1.0, 1.0) }
            
            withContext(Dispatchers.Main) {
                result.success(simulatedEmbeddings)
            }
        }
    }

    private fun embedTextBatch(call: MethodCall, result: Result) {
        val texts = call.argument<List<String>>("texts")
        
        if (texts == null) {
            result.success(listOf<List<Double>>())
            return
        }

        if (!embeddingModelLoaded) {
            result.success(listOf<List<Double>>())
            return
        }

        println("$TAG: Generating batch embeddings for ${texts.size} texts")

        // TODO: Batch embedding generation
        CoroutineScope(Dispatchers.IO).launch {
            delay(texts.size.toLong() * 100) // Simulate batch processing
            
            val batchEmbeddings = texts.map { 
                (0 until 512).map { Random.nextDouble(-1.0, 1.0) }
            }
            
            withContext(Dispatchers.Main) {
                result.success(batchEmbeddings)
            }
        }
    }

    // MARK: - Device Capabilities

    private fun getDeviceCapabilities(result: Result) {
        val context = activity ?: return result.success(null)
        
        val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val memInfo = ActivityManager.MemoryInfo()
        activityManager.getMemoryInfo(memInfo)
        
        val totalMemoryMB = (memInfo.totalMem / (1024 * 1024)).toInt()
        val availableMemoryMB = (memInfo.availMem / (1024 * 1024)).toInt()
        
        val capabilities = mapOf(
            "totalRamMB" to totalMemoryMB,
            "availableRamMB" to availableMemoryMB,
            "deviceModel" to "${Build.MANUFACTURER} ${Build.MODEL}",
            "osVersion" to "Android ${Build.VERSION.RELEASE}",
            "apiLevel" to Build.VERSION.SDK_INT
        )
        
        println("$TAG: Device capabilities - ${totalMemoryMB}MB RAM")
        result.success(capabilities)
    }

    // MARK: - Model Management

    private fun isModelReady(call: MethodCall, result: Result) {
        val modelType = call.argument<String>("modelType")
        
        val ready = when (modelType) {
            "chat" -> chatModelLoaded
            "vision" -> visionModelLoaded  
            "embedding" -> embeddingModelLoaded
            else -> false
        }
        
        result.success(ready)
    }

    private fun getModelLoadingProgress(call: MethodCall, result: Result) {
        val modelType = call.argument<String>("modelType")
        
        // Simulate loading progress
        val progress = when (modelType) {
            "chat" -> if (chatModelLoaded) 1.0 else 0.7
            "vision" -> if (visionModelLoaded) 1.0 else 0.5
            "embedding" -> if (embeddingModelLoaded) 1.0 else 0.9
            else -> 0.0
        }
        
        result.success(progress)
    }

    // MARK: - Runtime Management

    private fun switchRuntime(call: MethodCall, result: Result) {
        val runtime = call.argument<String>("runtime")
        
        if (runtime == null) {
            result.success(false)
            return
        }

        println("$TAG: Switching to runtime: $runtime")
        currentRuntime = runtime
        
        // TODO: Actually switch between llama.cpp and MLC runtimes
        result.success(true)
    }

    private fun getRuntimeInfo(result: Result) {
        val runtimeInfo = mapOf(
            "runtime" to currentRuntime,
            "version" to "stub-1.0.0",
            "supportedModels" to listOf(
                "qwen3-4b-instruct", 
                "qwen2.5-vl-3b", 
                "qwen3-embedding-0.6b"
            )
        )
        
        result.success(runtimeInfo)
    }

    // MARK: - Cleanup

    private fun dispose(result: Result) {
        println("$TAG: Disposing models and cleaning up resources")
        
        // Clean up llama.cpp resources
        try {
            llamaCleanup()
        } catch (e: Exception) {
            println("$TAG: Error during cleanup: ${e.message}")
        }
        
        chatModelLoaded = false
        visionModelLoaded = false
        embeddingModelLoaded = false
        
        result.success(null)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}

File: ARC MVP/EPI/ios/Runner/QwenBridge.swift
import Foundation
import Flutter
import UIKit

// Import the C++ wrapper
// Note: In a real implementation, this would be linked to the actual llama.cpp library

// C function declarations for llama.cpp integration
@_silgen_name("llama_init")
func llama_init(_ modelPath: UnsafePointer<CChar>) -> Int32

@_silgen_name("llama_generate")
func llama_generate(_ prompt: UnsafePointer<CChar>, 
                   _ temperature: Float, 
                   _ topP: Float, 
                   _ maxTokens: Int32) -> UnsafePointer<CChar>?

@_silgen_name("llama_cleanup")
func llama_cleanup()

@_silgen_name("llama_is_loaded")
func llama_is_loaded() -> Int32

// Stub implementation for Qwen models until llama.cpp integration is complete
// This maintains the API structure while providing development functionality

@objc class QwenBridge: NSObject, FlutterPlugin {
    // Model state tracking
    private var chatModelLoaded = false
    private var visionModelLoaded = false
    private var embeddingModelLoaded = false
    
    // Model configurations
    private var currentRuntime: String = "llamacpp"
    private var temperature: Float = 0.6
    private var topP: Float = 0.9
    private var maxTokens: Int = 256
    
    // Required FlutterPlugin method
    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "lumara_native", binaryMessenger: registrar.messenger())
        let instance = QwenBridge()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    // Custom register method for direct use
    static func register(with binaryMessenger: FlutterBinaryMessenger) {
        let channel = FlutterMethodChannel(name: "lumara_native", binaryMessenger: binaryMessenger)
        let instance = QwenBridge()
        channel.setMethodCallHandler(instance.handle)
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        // Chat model methods
        case "initChatModel":
            initChatModel(call: call, result: result)
        case "qwenText":
            qwenText(call: call, result: result)
            
        // Vision model methods
        case "initVisionModel":
            initVisionModel(call: call, result: result)
        case "qwenVision":
            qwenVision(call: call, result: result)
            
        // Embedding model methods
        case "initEmbeddingModel":
            initEmbeddingModel(call: call, result: result)
        case "embedText":
            embedText(call: call, result: result)
        case "embedTextBatch":
            embedTextBatch(call: call, result: result)
            
        // Device capabilities
        case "getDeviceCapabilities":
            getDeviceCapabilities(result: result)
            
        // Model management
        case "isModelReady":
            isModelReady(call: call, result: result)
        case "getModelLoadingProgress":
            getModelLoadingProgress(call: call, result: result)
            
        // Runtime management
        case "switchRuntime":
            switchRuntime(call: call, result: result)
        case "getRuntimeInfo":
            getRuntimeInfo(result: result)
            
        case "dispose":
            dispose(result: result)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - Chat Model Methods
    
    private func initChatModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let modelPath = args["modelPath"] as? String else {
            result(false)
            return
        }
        
        let temperature = args["temperature"] as? Double ?? 0.6
        let topP = args["top_p"] as? Double ?? 0.9
        let maxTokens = args["max_tokens"] as? Int ?? 256
        
        print("QwenBridge: Initializing chat model")
        print("  Model path: \(modelPath)")
        print("  Temperature: \(temperature)")
        print("  Top-P: \(topP)")
        print("  Max tokens: \(maxTokens)")
        
        // Store parameters for later use
        self.temperature = Float(temperature)
        self.topP = Float(topP)
        self.maxTokens = maxTokens
        
        // Initialize llama.cpp model
        DispatchQueue.global(qos: .userInitiated).async {
            print("QwenBridge: Calling native llama_init with path: \(modelPath)")
            let success = llama_init(modelPath)
            print("QwenBridge: Native llama_init returned: \(success)")
            
            DispatchQueue.main.async {
                if success == 1 {
                    self.chatModelLoaded = true
                    print("QwenBridge: Chat model loaded successfully via native C++")
                    result(true)
                } else {
                    print("QwenBridge: Failed to load chat model via native C++")
                    result(false)
                }
            }
        }
    }
    
    private func qwenText(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let prompt = args["prompt"] as? String else {
            result("")
            return
        }
        
        guard chatModelLoaded else {
            print("QwenBridge: Chat model not loaded")
            result("")
            return
        }
        
        print("QwenBridge: Generating text for prompt: \(prompt.prefix(50))...")
        
        // Generate response using llama.cpp
        DispatchQueue.global(qos: .userInitiated).async {
            print("QwenBridge: Calling native llama_generate...")
            print("QwenBridge: Prompt length: \(prompt.count) characters")
            
            let response = llama_generate(prompt, self.temperature, self.topP, Int32(self.maxTokens))
            
            let responseString: String
            if let cString = response {
                responseString = String(cString: cString)
                print("QwenBridge: Native C++ generated response: \(responseString.prefix(100))...")
            } else {
                print("QwenBridge: Native C++ returned null, using fallback")
                responseString = self.generateContextualResponse(from: prompt)
            }
            
            DispatchQueue.main.async {
                result(responseString)
            }
        }
    }
    
    private func generateContextualResponse(from prompt: String) -> String {
        // Extract information from the prompt
        let hasJournalEntries = prompt.contains("journal entries") || prompt.contains("Sample journal entry")
        let hasArcforms = prompt.contains("Arcform") || prompt.contains("arcforms")
        let hasPhaseInfo = prompt.contains("Discovery") || prompt.contains("phase")
        let isChat = prompt.contains("CONVERSATION:") || prompt.contains("User:")
        
        var response = ""
        
        if isChat {
            // Extract the user's latest message
            if let userMessageRange = prompt.range(of: "Please respond to the user's latest message: \"") {
                let afterQuote = prompt[userMessageRange.upperBound...]
                if let endQuote = afterQuote.firstIndex(of: "\"") {
                    let userMessage = String(afterQuote[..<endQuote])
                    response = "I understand you're asking about \"\(userMessage)\". "
                }
            }
        }
        
        // Add contextual analysis based on available data
        if hasJournalEntries {
            response += "Based on your recent journal entries, I can see patterns emerging in your daily experiences. "
        }
        
        if hasPhaseInfo {
            response += "You appear to be in the Discovery phase, which suggests you're exploring new ideas and possibilities. "
        }
        
        if hasArcforms {
            response += "Your Arcform data shows interesting insights about your current state. "
        }
        
        // Add a helpful suggestion
        if response.isEmpty {
            response = "I'm here to help you explore your thoughts and patterns. "
        }
        
        response += "What would you like to understand better about your recent experiences?"
        
        return response
    }
    
    // MARK: - Vision Model Methods
    
    private func initVisionModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let modelPath = args["modelPath"] as? String else {
            result(false)
            return
        }
        
        print("QwenBridge: Initializing vision model at \(modelPath)")
        
        // TODO: Initialize actual Qwen-VL model
        DispatchQueue.global().asyncAfter(deadline: .now() + 1.5) {
            self.visionModelLoaded = true
            DispatchQueue.main.async {
                result(true)
            }
        }
    }
    
    private func qwenVision(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let prompt = args["prompt"] as? String,
              let imageData = args["imageJpeg"] as? FlutterStandardTypedData else {
            result("")
            return
        }
        
        guard visionModelLoaded else {
            print("QwenBridge: Vision model not loaded")
            result("")
            return
        }
        
        let imageSize = imageData.data.count
        print("QwenBridge: Analyzing image (\(imageSize) bytes) with prompt: \(prompt)")
        
        // TODO: Call actual Qwen-VL model
        let simulatedVisionResponse = """
        I can see the image you've shared. This appears to be a photo related to your journal entry or personal experience. 
        
        Your question: "\(prompt)"
        
        *This is a stub response. The actual Qwen2.5-VL model will provide detailed image analysis and answer questions about visual content once the integration is complete.*
        """
        
        DispatchQueue.global().asyncAfter(deadline: .now() + 1.0) {
            DispatchQueue.main.async {
                result(simulatedVisionResponse)
            }
        }
    }
    
    // MARK: - Embedding Model Methods
    
    private func initEmbeddingModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let modelPath = args["modelPath"] as? String else {
            result(false)
            return
        }
        
        print("QwenBridge: Initializing embedding model at \(modelPath)")
        
        // TODO: Initialize actual Qwen3-Embedding model
        DispatchQueue.global().asyncAfter(deadline: .now() + 0.5) {
            self.embeddingModelLoaded = true
            DispatchQueue.main.async {
                result(true)
            }
        }
    }
    
    private func embedText(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let text = args["text"] as? String else {
            result([])
            return
        }
        
        guard embeddingModelLoaded else {
            print("QwenBridge: Embedding model not loaded")
            result([])
            return
        }
        
        print("QwenBridge: Generating embeddings for text: \(text.prefix(50))...")
        
        // TODO: Generate actual embeddings with Qwen3-Embedding
        // For now, return simulated 512-dimensional embeddings
        let simulatedEmbeddings = (0..<512).map { _ in Double.random(in: -1.0...1.0) }
        
        DispatchQueue.global().asyncAfter(deadline: .now() + 0.1) {
            DispatchQueue.main.async {
                result(simulatedEmbeddings)
            }
        }
    }
    
    private func embedTextBatch(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let texts = args["texts"] as? [String] else {
            result([])
            return
        }
        
        guard embeddingModelLoaded else {
            result([])
            return
        }
        
        print("QwenBridge: Generating batch embeddings for \(texts.count) texts")
        
        // TODO: Batch embedding generation
        let batchEmbeddings = texts.map { _ in
            (0..<512).map { _ in Double.random(in: -1.0...1.0) }
        }
        
        DispatchQueue.global().asyncAfter(deadline: .now() + Double(texts.count) * 0.1) {
            DispatchQueue.main.async {
                result(batchEmbeddings)
            }
        }
    }
    
    // MARK: - Device Capabilities
    
    private func getDeviceCapabilities(result: @escaping FlutterResult) {
        let processInfo = ProcessInfo.processInfo
        let totalMemory = processInfo.physicalMemory
        let totalMemoryMB = Int(totalMemory / (1024 * 1024))
        
        // Estimate available memory (conservative)
        let availableMemoryMB = Int(Double(totalMemoryMB) * 0.6)
        
        let capabilities: [String: Any] = [
            "totalRamMB": totalMemoryMB,
            "availableRamMB": availableMemoryMB,
            "deviceModel": getDeviceModel(),
            "osVersion": UIDevice.current.systemVersion
        ]
        
        print("QwenBridge: Device capabilities - \(totalMemoryMB)MB RAM")
        result(capabilities)
    }
    
    private func getDeviceModel() -> String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let identifier = machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8, value != 0 else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value)))
        }
        return identifier
    }
    
    // MARK: - Model Management
    
    private func isModelReady(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let modelType = args["modelType"] as? String else {
            result(false)
            return
        }
        
        let ready: Bool
        switch modelType {
        case "chat":
            // Check both our flag and the native C++ status
            ready = chatModelLoaded && (llama_is_loaded() == 1)
        case "vision":
            ready = visionModelLoaded
        case "embedding":
            ready = embeddingModelLoaded
        default:
            ready = false
        }
        
        result(ready)
    }
    
    private func getModelLoadingProgress(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let modelType = args["modelType"] as? String else {
            result(0.0)
            return
        }
        
        // Simulate loading progress
        let progress: Double
        switch modelType {
        case "chat":
            progress = chatModelLoaded ? 1.0 : 0.7
        case "vision":
            progress = visionModelLoaded ? 1.0 : 0.5
        case "embedding":
            progress = embeddingModelLoaded ? 1.0 : 0.9
        default:
            progress = 0.0
        }
        
        result(progress)
    }
    
    // MARK: - Runtime Management
    
    private func switchRuntime(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let runtime = args["runtime"] as? String else {
            result(false)
            return
        }
        
        print("QwenBridge: Switching to runtime: \(runtime)")
        currentRuntime = runtime
        
        // TODO: Actually switch between llama.cpp and MLC runtimes
        result(true)
    }
    
    private func getRuntimeInfo(result: @escaping FlutterResult) {
        let runtimeInfo: [String: Any] = [
            "runtime": currentRuntime,
            "version": "stub-1.0.0",
            "supportedModels": ["qwen3-4b-instruct", "qwen2.5-vl-3b", "qwen3-embedding-0.6b"]
        ]
        
        result(runtimeInfo)
    }
    
    // MARK: - Cleanup
    
    private func dispose(result: @escaping FlutterResult) {
        print("QwenBridge: Disposing models and cleaning up resources")
        
        // Clean up llama.cpp resources
        llama_cleanup()
        
        chatModelLoaded = false
        visionModelLoaded = false
        embeddingModelLoaded = false
        
        print("QwenBridge: Cleanup complete")
        result(nil)
    }
}